File: cpp/bfs.cpp
// BFS benchmark
// Input: graph file (n m, then m edges u v)
// Output: read=Xms compute=Yms checksum=C
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <vector>

using Clock = std::chrono::steady_clock;

static double ms_since(Clock::time_point t0) {
  return std::chrono::duration<double, std::milli>(Clock::now() - t0).count();
}

int main(int argc, char **argv) {
  if (argc < 2) {
    fprintf(stderr, "Usage: %s <graph_file>\n", argv[0]);
    return 1;
  }

  // ── Read ─────────────────────────────────────────────────────────────────
  auto t0 = Clock::now();

  FILE *f = fopen(argv[1], "r");
  if (!f) {
    perror("fopen");
    return 1;
  }

  uint32_t n, m;
  fscanf(f, "%u %u", &n, &m);

  // CSR-style adjacency: offsets + flat neighbor array
  // First pass: count degrees
  std::vector<uint32_t> deg(n, 0);
  std::vector<uint32_t> eu(m);
  std::vector<uint32_t> ev(m);
  for (uint32_t i = 0; i < m; i++) {
    fscanf(f, "%u %u", &eu.at(i), &ev.at(i));
    deg.at(eu.at(i))++;
    deg.at(ev.at(i))++;
  }
  fclose(f);

  // Build adjacency arrays
  std::vector<uint32_t> offset(n + 1);
  offset.at(0) = 0;
  for (uint32_t i = 0; i < n; i++)
    offset.at(i + 1) = offset.at(i) + deg.at(i);
  uint32_t total_edges = offset.at(n);
  std::vector<uint32_t> adj(total_edges);
  std::vector<uint32_t> pos(n, 0);
  for (uint32_t i = 0; i < m; i++) {
    uint32_t u = eu.at(i), v = ev.at(i);
    adj.at(offset.at(u) + pos.at(u)++) = v;
    adj.at(offset.at(v) + pos.at(v)++) = u;
  }

  double read_ms = ms_since(t0);

  // ── Compute ──────────────────────────────────────────────────────────────
  auto t1 = Clock::now();

  // BFS from vertex 0
  std::vector<uint32_t> visited(n, 0);
  std::vector<uint32_t> queue(n);
  uint32_t qhead = 0, qtail = 0;

  visited.at(0) = 1;
  queue.at(qtail++) = 0;

  int64_t dist_sum = 0;
  std::vector<int64_t> dist(n, 0);

  while (qhead < qtail) {
    uint32_t v = queue.at(qhead++);
    for (uint32_t i = offset.at(v); i < offset.at(v + 1); i++) {
      uint32_t w = adj.at(i);
      if (!visited.at(w)) {
        visited.at(w) = 1;
        dist.at(w) = dist.at(v) + 1;
        dist_sum += dist.at(w);
        queue.at(qtail++) = w;
      }
    }
  }

  double compute_ms = ms_since(t1);

  // ── Output ───────────────────────────────────────────────────────────────
  printf("read=%.1fms compute=%.1fms checksum=%lld visited=%u\n", read_ms,
         compute_ms, (long long)dist_sum, qtail);

  return 0;
}

File: rust/src/bin/bfs.rs
// BFS benchmark
use std::env;
use std::fs;
use std::time::Instant;

#[inline(always)]
fn parse_usize(bytes: &[u8]) -> usize {
    let mut n: usize = 0;
    for &b in bytes {
        n = n * 10 + (b - b'0') as usize;
    }
    n
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <graph_file>", args[0]);
        std::process::exit(1);
    }

    // ── Read ─────────────────────────────────────────────────────────────────
    let t0 = Instant::now();

    let bytes = fs::read(&args[1]).unwrap();
    let mut iter = bytes
        .split(|&b| b == b' ' || b == b'\n' || b == b'\r' || b == b'\t')
        .filter(|s| !s.is_empty());
    let n: usize = parse_usize(iter.next().unwrap());
    let m: usize = parse_usize(iter.next().unwrap());

    // Build CSR adjacency
    let mut deg = vec![0u32; n];
    let mut eu = Vec::with_capacity(m);
    let mut ev = Vec::with_capacity(m);
    for _ in 0..m {
        let u: usize = parse_usize(iter.next().unwrap());
        let v: usize = parse_usize(iter.next().unwrap());
        deg[u] += 1;
        deg[v] += 1;
        eu.push(u);
        ev.push(v);
    }

    let mut offset = vec![0u32; n + 1];
    for i in 0..n {
        offset[i + 1] = offset[i] + deg[i];
    }
    let total = offset[n] as usize;
    let mut adj = vec![0u32; total];
    let mut pos = vec![0u32; n];
    for i in 0..m {
        let u = eu[i];
        let v = ev[i];
        let pu = offset[u] + pos[u];
        adj[pu as usize] = v as u32;
        pos[u] += 1;
        let pv = offset[v] + pos[v];
        adj[pv as usize] = u as u32;
        pos[v] += 1;
    }

    let read_ms = t0.elapsed().as_secs_f64() * 1000.0;

    // ── Compute ──────────────────────────────────────────────────────────────
    let t1 = Instant::now();

    let mut visited = vec![false; n];
    let mut queue = Vec::with_capacity(n);
    let mut dist = vec![0i64; n];

    visited[0] = true;
    queue.push(0u32);
    let mut qhead = 0usize;
    let mut dist_sum: i64 = 0;

    while qhead < queue.len() {
        let v = queue[qhead] as usize;
        qhead += 1;
        let lo = offset[v] as usize;
        let hi = offset[v + 1] as usize;
        for &w_raw in &adj[lo..hi] {
            let w = w_raw as usize;
            if !visited[w] {
                visited[w] = true;
                let d = dist[v] + 1;
                dist[w] = d;
                dist_sum += d;
                queue.push(w as u32);
            }
        }
    }

    let compute_ms = t1.elapsed().as_secs_f64() * 1000.0;

    println!(
        "read={:.1}ms compute={:.1}ms checksum={} visited={}",
        read_ms,
        compute_ms,
        dist_sum,
        queue.len()
    );
}
